toolchain/sem_ir/inst_profile.cpp

```diff
@@ -4,110 +4,115 @@
-static auto IntProfileArgFunction(llvm::FoldingSetNodeID& id,
-                                  const File& sem_ir, int32_t arg) -> void {
-  sem_ir.ints().Get(IntId(arg)).Profile(id);
+static auto ProfileArg(llvm::FoldingSetNodeID& id, const File& sem_ir,
+                       IntId int_id) -> void {
+  sem_ir.ints().Get(int_id).Profile(id);
 }
 
 // Profiling for real number IDs.
-static auto RealProfileArgFunction(llvm::FoldingSetNodeID& id,
-                                   const File& sem_ir, int32_t arg) -> void {
-  const auto& real = sem_ir.reals().Get(RealId(arg));
+static auto ProfileArg(llvm::FoldingSetNodeID& id, const File& sem_ir,
+                       RealId real_id) -> void {
+  const auto& real = sem_ir.reals().Get(real_id);
   // TODO: Profile the value rather than the syntactic form.
   real.mantissa.Profile(id);
   real.exponent.Profile(id);
   id.AddBoolean(real.is_decimal);
 }
 
-// Selects the function to use to profile argument N of instruction InstT. We
-// compute this in advance so that we can reuse the profiling code for all
-// instructions that are profiled in the same way. For example, all instructions
-// that take two IDs that are profiled by value use the same profiling code,
-// namely `ProfileArgs<DefaultProfileArgFunction, DefaultProfileArgFunction>`.
+// Wrap is_invocable_v to avoid a compile error on an unused lambda operand.
+template <typename... T, typename F>
+constexpr auto IsInvocable(F /*f*/) -> bool {
+  return std::is_invocable_v<F, T...>;
+}
+
+// Selects the argument type to use to profile argument N of instruction InstT.
+// We map as many types as we can to `int32_t` so that we can reuse the
+// profiling code for all instructions that are profiled in the same way. For
+// example, all instructions that take two IDs that are profiled by value use
+// the same profiling code.
 template <typename InstT, int N>
-static constexpr auto SelectProfileArgFunction() -> ProfileArgFunction* {
+static auto ComputeProfileArgType() -> auto {
   if constexpr (N >= InstLikeTypeInfo<InstT>::NumArgs) {
     // This argument is not used by this instruction; don't profile it.
-    return NullProfileArgFunction;
+    return NoArg{0};
   } else {
     using ArgT = typename InstLikeTypeInfo<InstT>::template ArgType<N>;
-    if constexpr (std::is_same_v<ArgT, InstBlockId>) {
-      return InstBlockProfileArgFunction;
-    } else if constexpr (std::is_same_v<ArgT, TypeBlockId>) {
-      return TypeBlockProfileArgFunction;
-    } else if constexpr (std::is_same_v<ArgT, IntId>) {
-      return IntProfileArgFunction;
-    } else if constexpr (std::is_same_v<ArgT, RealId>) {
-      return RealProfileArgFunction;
+    if constexpr (IsInvocable<ArgT>(
+                      [](auto&& arg)
+                          -> decltype(ProfileArg(
+                              std::declval<llvm::FoldingSetNodeID&>(),
+                              std::declval<const File&>(), arg)) {})) {
+      return std::declval<ArgT>();
```

Comment:
I think it's better to have an explicit list of the ID types that use the `int32_t` profiling rather than using that profiling automatically for newly-added ID types, so that explicit action is required when adding a new ID type to choose the profiling behavior. (In particular, while making the list I noticed that we were using `int32_t` profiling for `BindNameId` and for `StringLiteralValueId`, both of which seem questionable.)